daemon/cmd/fqdn.go

```diff
@@ -586,6 +625,39 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 	return nil
 }
 
+// getMutexesForResponseIPs returns a slice of indices for accessing the
+// mutexes in dnsProxyContext. There's a many-to-one mapping from IP to mutex,
```

Comment:
Should this be operating on `dnsProxyContext` instead of `Daemon` then?

Also nit: `getMutexIndicesForResponseIPs()`? Or it could even just grab the locks rather than relying on the calling logic to handle that aspect, since it's so crucial to avoiding deadlocks. :shrug: That's a bit more cosmetic than anything though.

Reply:
I've updated the function to be operator on `dnsProxyContext` instead. 

Yeah, `notifyOnDNSMsg()` is starting to get quite large and difficult to read, a refactor is definitely in order. I agree that with the big comment and all the little details is far too in the weeds. Hopefully we can improve this with a refactor at some point.